lib/ecto/query.ex

```diff
@@ -1234,8 +1236,9 @@ defmodule Ecto.Query do
   invalid queries. If you are running into such scenarios, your best
   option is to use a fragment as your CTE.
   '''
-  defmacro with_cte(query, name, as: with_query) do
-    Builder.CTE.build(query, name, with_query, __CALLER__)
+  defmacro with_cte(query, name, [{:as, with_query} | opts]) do
```

Comment:
The `as` keyword now needs to be the first list item. Is that desired?

Reply:
Based in the docs from https://www.postgresql.org/docs/15/queries-with.html (section 7.8) the behaviour is a bit more complex than it may seem at first:
- Not specifying either `MATERIALIZED` or `NOT MATERIALIZED` will leave up to postgres whether to materialize
- Specifying overrides the postgres selection, which can bring performance improvements.
So we would need to support both options.

Reply:
Shall we code it with `:not_materialized | :materialized | nil` internally? `nil` would leave it up to postgres whether to materialize.

Reply:
I'm not sure, cc: @josevalim, @greg-rychlewski

Reply:
Can't it be like this

```elixir
with_cte(name, query, opts)

## options
  `:as` - the CTE query itself or a fragment
  `:materialize - a boolean indicating whether the CTE should be materialized. if blank, the database's default behaviour will be used
```  

Also I don't think there's a need to force the keyword list order. You can just make sure they specify `:as` and raise if not.

Reply:
:+1: for @greg-rychlewski's proposal!